### PR TITLE
GEM-62 | Add registration

### PR DESCRIPTION
**Pull request by pwgacek**

### CODE REVIEW

File: src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        val registrationRequest = createRegistrationRequest()
+
+        // when
+        val response = service.register(registrationRequest)
+
+        // then
+        response shouldHaveHttpStatus CREATED
+    }
+
+    should("return validation exception when email is blank") {
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/controller/AuthControllerIT.kt]: could we change that cases to single test with data for example 
```kotlin
            context("return validation exception cause:") {
                withData(
                        nameFn = { it.first },
                        Pair("Email can not be blank", createRegistrationRequest(email = "")),
                        Pair("Wrong email format", createRegistrationRequest(email = "email")),
                        Pair("Password can not be blank", createRegistrationRequest(password = "")),
                        Pair("Minimum password length is 8", createRegistrationRequest(password = "pswd")),
                        Pair("Maximum password length is 30", createRegistrationRequest(password = "passwordpasswordpasswordpassword")),
                        Pair("Password must contain at least one lowercase letter", createRegistrationRequest(password = "PASSWORD")),
                        Pair("Password must contain at least one uppercase letter", createRegistrationRequest(password = "password")),
                        Pair("Password must contain at least one digit", createRegistrationRequest(password = "password")),
                        Pair("Password must contain at least one special character among @#\$%^&+=!", createRegistrationRequest(password = "password")),
                ) { (expectedMessage, registrationRequest) ->
                    // when
                    val response = service.register(registrationRequest)

                    // then
                    response shouldHaveHttpStatus BAD_REQUEST
                    response shouldHaveValidationError expectedMessage
                }
            }
```

File: src/main/kotlin/pl/edu/agh/gem/config/AuthConfig.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+@Configuration
+@EnableWebSecurity
+class AuthConfig {
+    @Bean
+    fun encoder(): PasswordEncoder {
+        return BCryptPasswordEncoder()
+    }
+
+    @Bean
+    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/config/AuthConfig.kt:20]: do we still need spring security configuration if we want to permit any request ?

File: src/main/kotlin/pl/edu/agh/gem/external/client/RestEmailSenderClient.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @Qualifier("EmailSenderClientRestTemplate") val restTemplate: RestTemplate,
+    private val emailSenderClientProperties: EmailSenderClientProperties,
+) : EmailSenderClient {
+
+    @Retry(name = "default")
+    override fun sendVerificationEmail(verificationEmailDetails: VerificationEmailDetails) {
+        try {
+            restTemplate.exchange(
+                resolveVerificationAddress(),
+                HttpMethod.POST,
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/client/RestEmailSenderClient.kt]: static import

File: src/main/kotlin/pl/edu/agh/gem/external/controller/ApiExceptionHandler.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                    .withDetails(error.field)
+                    .withUserMessage(error.defaultMessage)
+                    .withMessage(error.defaultMessage)
+            }
+        return SimpleErrorsHolder(errors).apply {
+            jacksonObjectMapper().writeValueAsString(this)
+        }
+    }
+
+    private fun handleError(exception: Exception): SimpleErrorsHolder {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/controller/ApiExceptionHandler.kt]: i have move that to lib

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/MongoNotVerifiedUserRepository.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+@Repository
+class MongoNotVerifiedUserRepository(
+    private val mongo: MongoTemplate,
+) : NotVerifiedUserRepository {
+    override fun create(notVerifiedUser: NotVerifiedUser): NotVerifiedUser {
+        return mongo.insert(notVerifiedUser.toEntity()).toDomain()
+    }
+
+    override fun existByEmail(email: String): Boolean {
+        val query = Query().addCriteria(where("email").`is`(email))
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/MongoNotVerifiedUserRepository.kt]: could you use "NotVerifiedUserEntity::email.name" to not use simple string?

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import pl.edu.agh.gem.internal.model.auth.NotVerifiedUser
+import java.time.LocalDateTime
+
+@Document("not-verified")
+data class NotVerifiedUserEntity(
+    @Id
+    val id: String,
+    val email: String,
+    val password: String,
+    @Indexed(name = "ttl_index", expireAfter = "P30D")
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt]: do we need to use that name, or sth else or auto name ?

File: src/main/kotlin/pl/edu/agh/gem/internal/service/AuthService.kt [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+@Service
+class AuthService(
+    private val notVerifiedUserRepository: NotVerifiedUserRepository,
+    private val verifiedUserRepository: VerifiedUserRepository,
+    private val senderClient: EmailSenderClient,
+) {
+
+    fun create(notVerifiedUser: NotVerifiedUser) {
+        if (notVerifiedUserRepository.existByEmail(notVerifiedUser.email) || verifiedUserRepository.existByEmail(notVerifiedUser.email)) {
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/AuthService.kt]: could we extract that check to other private function ?

File: src/main/kotlin/pl/edu/agh/gem/internal/service/AuthService.kt [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            .collect(Collectors.joining(""))
+    }
+
+    companion object {
+        private const val CODE_LENGTH = 6L
+        private const val RANDOM_NUMBER_BOUND = 10
+    }
+}
+
+class DuplicateEmailException : RuntimeException("Email is already taken")
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/internal/service/AuthService.kt]: could we add duplicated email inside that message ?

**Comment by lawos98** [comment]
dont maching name and PR title

File: src/main/kotlin/pl/edu/agh/gem/config/AuthConfig.kt (hunk 2 above)
Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/config/AuthConfig.kt:20]: yes, its a side effect of using spring security for authentication. Without this filter all request would be blocked

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt (hunk 6 above)
Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt]: we don't. will remove it and then mongo will set default name

File: src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import java.time.LocalDateTime
+
+@Document("not-verified")
+data class NotVerifiedUserEntity(
+    @Id
+    val id: String,
+    val email: String,
+    val password: String,
+    @Indexed(expireAfter = "P30D")
+    val createdAt: LocalDateTime,
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt:16]: do we want to change that to Instant?

Comment (pwgacek) [reply src/main/kotlin/pl/edu/agh/gem/external/persistence/NotVerifiedUserEntity.kt:16]: will add in last pr

File: src/integration/kotlin/pl/edu/agh/gem/integration/ability/EmailSenderAbility.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import com.github.tomakehurst.wiremock.client.WireMock.post
+import com.github.tomakehurst.wiremock.client.WireMock.urlMatching
+import org.springframework.http.HttpStatus
+import org.springframework.http.HttpStatusCode
+import pl.edu.agh.gem.headers.HeadersTestUtils.withAppContentType
+import pl.edu.agh.gem.integration.environment.ProjectConfig
+
+private const val VERIFICATION_URL = "/internal/verification"
+
+fun stubEmailSenderVerification(statusCode: HttpStatusCode = HttpStatus.OK) {
```
Comment (lawos98) [inline src/integration/kotlin/pl/edu/agh/gem/integration/ability/EmailSenderAbility.kt:13]: static but ok

File: src/main/kotlin/pl/edu/agh/gem/external/dto/auth/RegistrationRequest.kt [hunk 11]
```diff
@@ -0,0 +1,24 @@
+package pl.edu.agh.gem.external.dto.auth
+
+import jakarta.validation.constraints.NotBlank
+import jakarta.validation.constraints.Pattern
+import jakarta.validation.constraints.Size
+
+data class RegistrationRequest(
+    @field:NotBlank(message = "Email can not be blank")
+    @field:Pattern(
```
Comment (lawos98) [inline src/main/kotlin/pl/edu/agh/gem/external/dto/auth/RegistrationRequest.kt:9]: could we use instead @Email ?

**Review by lawos98 — APPROVED** [review]

**Pull request merged**